mujmap: add module

elizagamedev opened the pull request:
mujmap is a tool which synchronizes mail between a mail server and
notmuch via JMAP. It's very similar to lieer, so I heavily based the
implementation of the notmuch module on lieer's. I did not include an
equivalent to lieer's periodic synchronization service, however, because
I plan to soon introduce a daemon mode to mujmap.

https://github.com/elizagamedev/mujmap

This isn't quite ready for merge yet, as I'm still waiting on
NixOS/nixpkgs#172648. However, this is my first module contribution to
home-manager, and I thought it was worth it to receive early feedback.

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

teto commented:
Great job ! 

My generated mujmap.toml contained `password_command = ["/nix/store/acfizib24wk6hq20qg5x9f3n8398q6lh-pass-show/bin/pass-show", "perso/fastmail_mc"]` which is apparently not supported by mujmap (it expects a string).

Also to setup the module I had to set mujmap.fqdn: in modules/accounts/email.nix there is a setting called "flavor" which lists "fastmail.com" and could be used to autoset fqdn

elizagamedev commented:
Pushed a commit to fix both of these issues.

teto commented:
@elizagamedev CI fails for mundane reasons (can't find you as a maintainer), could you fix it ? other than that it looks good we have tests, integration with the mail backend, a pretty good first module. I will do a better review after testing/CI passes

elizagamedev commented:
@teto I'm not actually sure how to fix CI, since I added myself to the maintainers list but it's still failing. Is there another place I missed?

teto commented:
ha yeah I've tried it locally and the maintainer infra is complex. @rycee any idea ?

elizagamedev commented:
I used snake case so that it could directly be serialized to the config file like the lieer module does, but if it is preferred to have them as camel case it should not be too hard to translate the names before serialization.

teto commented:
dont do it yet as I am unsure, waiting confirmation from the HM gods :)

teto commented:
while doing a nix
```
error: store path 'm34v88fkaklilvsvv75vld6q384zqyra-mujmap-MYUSERNAME@fastmail.com.toml' contains illegal character '@'

       … while evaluating 'check'
```
Once could do
`lib.replaceStrings ["@"] ["a"] "t@to" `
or call sanitizeDerivationName  from lib/strings.nix

teto commented:
so using the module I get 
```
$ mujmap sync -v
[2022-06-16T22:19:30Z WARN  mujmap::sync] Could not read mujmap state file `./mujmap.state.json': No such file or directory (os error 2)
Retrieving metadata... (11157 possibly changed)
Applying changes to notmuch database... (11157 new, 11157 changed, 0 destroyed)
error: Could not sync mail: Could not add new local email: Something went wrong trying to read or write a file
```
the error is not verbose enough for a proper diagnostic. I still have my maildir on another drive (ext4), can it be it ?
```
ls -l ~/maildir
lrwxrwxrwx 1 teto users 16 2021-10-02  /home/teto/maildir -> /mnt/ext/maildir
```

elizagamedev commented:
> Once could do `lib.replaceStrings ["@"] ["a"] "t@to" ` or call sanitizeDerivationName from lib/strings.nix

Done.

> the error is not verbose enough for a proper diagnostic. I still have my maildir on another drive (ext4), can it be it ?

This is possible. I added some new config options for v2.0 which I forgot to add to this module which solve that problem (the [troubleshooting section of the readme](https://github.com/elizagamedev/mujmap/blob/main/README.md#troubleshooting) now addresses this). They have been added to the module. Does messing with those settings help?

I also tweaked `fqdnXorSessionUrlAssertions`, since v2.0 can now auto-detect the fqdn from the domain part of the username. Now it just checks to make sure you don't specify both.

elizagamedev commented:
> I get the same error `error: store path 'm34v88fkaklilvsvv75vld6q384zqyra-mujmap-MYUSERNAME@fastmail.com.toml' contains illegal character '@'` with your change. Actually I think sanitizeDerivationName is broken, I've discovered and had the same problem recently. For now `lib.replaceStrings ["@"] ["a"] "t@to" ` is probably enough.

I changed it to replace "`@`" with "`_at_`" since that seems pretty unique. I don't know why it was working for me and not you, maybe a difference in nix version?

> I've found the piece that recommands camelCase https://nix-community.github.io/home-manager/index.html#sec-code-style

It actually looks like there's an exception to this in the [manual](https://nix-community.github.io/home-manager/). (Section 5.2.3, Add only valuable options)

> If the software uses a structured configuration format like a JSON, YAML, INI, TOML, or even a plain list of key/value pairs then consider using a settings option as described in [Nix RFC 42](https://github.com/Infinisil/rfcs/blob/config-option/rfcs/0042-config-option.md).

The RFC linked is what lieer does, and because this is modeled after lieer's module, I inherited this way of doing things without realizing it's the "new" way. So changing the module to use camelCase would actually be going backwards. Perhaps the code style sction you linked ought to be amended to point this exception out to prevent future confusion.

That said, the "essential-ness" of some of the config options I explicitly included is a bit questionable. I removed the following explicit options:

- `convert_dos_to_unix`
- `timeout`
- `retries`
- `concurrent_downloads`

rycee commented on modules/lib/maintainers.nix:
```diff
@@ -281,4 +281,10 @@
       fingerprint = "D5D6 FD1F 0D9A 3284 FB9B  C26D 3F98 EC7E C2B8 7ED1";
     }];
   };
+  elizagamedev = {
+    email = "eliza@eliza.sh";
+    github = "elizagamedev";
+    githubId = 4576666;
+    name = "Eliza Velasquez";
+  };
```
You can remove this since you are already listed as a maintainer in Nixpkgs.
```suggestion
```

rycee commented on modules/programs/mujmap.nix:
```diff
@@ -0,0 +1,312 @@
+    assertion = a.mujmap.settings.fqdn == null || a.mujmap.settings.session_url
+      == null;
+    message =
+      "For `accounts.email.accounts.${a.name}', `notmuch.settings' may not specify both `fqdn' and `session_url'";
```
Perhaps you mean
```suggestion
      "For `accounts.email.accounts.${a.name}', `mujsmap.settings' may not specify both `fqdn' and `session_url'";
```

rycee commented on modules/programs/mujmap.nix:
```diff
@@ -0,0 +1,312 @@
+    map (name: "accounts.email.accounts.${name}.notmuch.enable = true;")
+    missingNotmuchAccounts;
+
+  fqdnForFlavor = flavor:
+    if flavor == "fastmail.com" then "fastmail.com" else null;
```
It may be better to add support for jmap in `accounts/email.nix` and then add an entry for jmap in the

```nix
mkIf (config.flavor == "fastmail.com") { … }
```

block. Then that configuration could be used in other modules that support JMAP.

teto replied:
that's a good idea, the same way we have an imapModule/smtpModule, we could have a jmapModule.

elizagamedev commented:
Thank you for your feedback, @rycee. I added a JMAP config option and modified the mujmap module to use it. Let me know if you want additional changes or if this is good to merge.

rycee commented:
Nice! I've squashed and merged to master now. Thanks for the contribution!
